src/test/spec/unified_runner/test_file.rs

```diff
@@ -0,0 +1,379 @@
+
+    let mut default_uri_parts = DEFAULT_URI.split('?');
+
+    let mut uri = String::from(default_uri_parts.next().unwrap());
+    // A connection string has two slashes before the host list and one slash before the auth db
+    // name. If an auth db name is not provided the latter slash might not be present, so it needs
+    // to be added manually.
+    if uri.chars().filter(|c| *c == '/').count() < 3 {
+        uri.push('/');
+    }
+    uri.push('?');
+
+    if let Some(options) = default_uri_parts.next() {
+        let options = options.split('&');
+        for option in options {
+            let key = option.split('=').next().unwrap();
+            // The provided URI options should override any existing options in the connection
+            // string.
+            if !uri_options.contains_key(key) {
+                uri.push_str(option);
+                uri.push('&');
+            }
+        }
+    }
+
+    for (key, value) in uri_options {
+        let value = value.to_string();
+        // to_string() wraps quotations around Bson strings
+        let value = value.trim_start_matches('\"').trim_end_matches('\"');
+        uri.push_str(&format!("{}={}&", &key, value));
+    }
+
+    // remove the trailing '&' from the URI (or '?' if no options are present)
+    uri.pop();
+
+    Ok(uri)
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Database {
+    pub id: String,
+    pub client: String,
+    pub database_name: String,
+    pub database_options: Option<CollectionOrDatabaseOptions>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Collection {
+    pub id: String,
+    pub database: String,
+    pub collection_name: String,
+    pub collection_options: Option<CollectionOrDatabaseOptions>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Session {
+    pub id: String,
+    pub client: String,
+    pub session_options: Option<Document>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Bucket {
+    pub id: String,
+    pub database: String,
+    pub bucket_options: Option<Document>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Stream {
+    pub id: String,
+    pub hex_bytes: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct CollectionOrDatabaseOptions {
+    // TODO properly implement Deserialize for ReadConcern
+    pub read_concern: Option<ReadConcern>,
+    #[serde(rename = "readPreference")]
+    pub selection_criteria: Option<SelectionCriteria>,
+    pub write_concern: Option<WriteConcern>,
+}
+
+impl CollectionOrDatabaseOptions {
+    pub fn as_database_options(&self) -> DatabaseOptions {
+        let mut database_options = DatabaseOptions::builder().build();
```

Comment:
Is there a reason you didn't just use the builder API here? i.e.

```rust
DatabaseOptions::builder()
    .read_concern(self.read_concern.clone())
    .selection_critiera(self.selection_criteria.clone())
    .write_concern(self.write_concern.clone())
    .build()
```

Since this is within the driver, you could also just use a struct literal if you prefer that.